dmoj/executors/filesystem_policies.py

```diff
@@ -0,0 +1,109 @@
+from typing import List, Union
+
+
+class ACCESS_MODE:
+    NONE = 0
+    EXACT = 1
+    RECURSIVE = 2
+
+
+class Dir:
+    def __init__(self):
+        self.access_mode = ACCESS_MODE.NONE
+
+        self.map = {}
+
+
+class File:
+    pass
+
+
+class ExactFile:
+    def __init__(self, path : str):
+        self.path = path
+
+
+class ExactDir:
+    access_mode = ACCESS_MODE.EXACT
+
+    def __init__(self, path : str):
+        self.path = path
+
+
+class RecursiveDir:
+    access_mode = ACCESS_MODE.RECURSIVE
+
+    def __init__(self, path : str):
+        self.path = path
+
+
+Rule = Union[ExactFile, ExactDir, RecursiveDir]
+
+
+class FilesystemPolicy:
+    def __init__(self, rules : List[Rule]):
+        self.rules = rules
+
+        self.built = False
+
+    def __add__(self, other):
+        assert not self.built and not other.built, "Refusing to merge built policies: merge before building"
+
+        return FilesystemPolicy(self.rules + other.rules)
+
+    def build(self):
+        self.root = Dir()
+
+        for rule in self.rules:
+            self.add_rule(rule)
+
+        self.built = True
+
+    def add_rule(self, rule : Rule):
```

Comment:
Will this method ever be called by the user outside of `build()`? If not, I think we usually prefix with `_` for clarity. If it will, perhaps we should add an assertion that `build()` was called so that you don't get a confusing `AttributeError` when `self.root` doesn't exist.

Reply:
I think in reality, the flow is "compose FilesystemPolicy objects into `fs` and then have the Tracer build it", in which case `build` could be prefixed with `_` as well.